sdk/metric/aggregator/exponential/structure/exponential.go

```diff
@@ -0,0 +1,657 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package structure // import "go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/structure"
+
+import (
+	"fmt"
+	"math/bits"
+
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping"
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping/exponent"
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping/logarithm"
+)
+
+type (
+	// Histogram observes counts observations in exponentially-spaced
+	// buckets.  It is configured with a maximum scale factor
+	// which determines resolution.  Scale is automatically
+	// adjusted to accommodate the range of input data.
+	//
+	// Note that the generic type `N` determines the type of the
+	// Sum, Min, and Max fields.  Bucket boundaries are handled in
+	// floating point regardless of the type of N.
+	Histogram[N ValueType] struct {
+		// maxSize is the maximum capacity of the positive and
+		// negative ranges.  it is set by Init(), preserved by
+		// Copy and Move.
+		maxSize int32
+
+		// sum is the sum of all Updates reflected in the
+		// aggregator.  It has the same type number as the
+		// corresponding sdkinstrument.Descriptor.
+		sum N
+		// count is incremented by 1 per Update.
+		count uint64
+		// zeroCount is incremented by 1 when the measured
+		// value is exactly 0.
+		zeroCount uint64
+		// min is set when count > 0
+		min N
+		// max is set when count > 0
+		max N
+		// positive holds the positive values
+		positive Buckets
+		// negative holds the negative values in these buckets
+		// by their absolute value.
+		negative Buckets
+		// mapping corresponds to the current scale, is shared
+		// by both positive and negative ranges.
+		mapping mapping.Mapping
+	}
+
+	// Buckets stores counts for measurement values in the range
+	// (0, +Inf).
+	Buckets struct {
+		// backing is a slice of nil, []uint8, []uint16, []uint32, or []uint64
+		backing bucketsBacking
+
+		// indexBase is index of the 0th position in the
+		// backing array, i.e., backing[0] is the count associated with
+		// indexBase which is in [indexStart, indexEnd]
+		indexBase int32
```

Comment:
should this and other indices be uint32?

Reply:
I've added a comment saying why this is signed. Values less than or equal to 1.0 have negative index values.